File: test/test_lexer.py
import pytest
from tinyc_lexer import Lexer
from tinyc_token import *
from util import make_tempfile_with_content, delete_tempfile

MODE = 0

def test_lex_simple():
    s = "i = 7;"
    c_file = make_tempfile_with_content(s)

    lexer = Lexer()
    lexer.open_file(c_file)

    tokens = [lexer.next_token()]
    while not tokens[-1].ttype == TOKEN_EOF:
        tokens.append(lexer.next_token())
        
    assert tokens[0].ttype == TOKEN_IDENT
    assert tokens[1].ttype == TOKEN_EQUAL
    assert tokens[2].ttype == TOKEN_INT
    assert tokens[3].ttype == TOKEN_SEMICOL

    delete_tempfile(c_file)

def test_lex_simple_wrong():
    s = "d = 4.2"
    c_file = make_tempfile_with_content(s)

    lexer = Lexer()
    lexer.open_file(c_file)

    lexer.next_token() # read d
    lexer.next_token() # read =

    with pytest.raises(Exception):
        lexer.next_token() # read 4.2

    delete_tempfile(c_file)

File: c/test_lexer.c
#include "lexer.h"
#include <assert.h>
#include <stdio.h>


int test_lexer_simple(void) {
    set_text("int i = 7;");
    set_pos(0);

    struct Token t0 = next_token();
    assert(t0.type == TOKEN_INT);
    
    struct Token t1 = next_token();
    struct Token t2 = next_token();
    struct Token t3 = next_token();
    struct Token t4 = next_token();
    struct Token t5 = next_token();
    struct Token t6 = next_token();
    struct Token t7 = next_token();
    struct Token t8 = next_token();

    printf("fdg\n");
    return 0;
}

int main(void) {

    test_lexer_simple();

    return 0;
}